client/core/trade.go

```diff
@@ -334,132 +329,134 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
-		isCancel := t.cancel != nil && msgMatch.Address == ""
-		if isCancel {
-			match.setStatus(order.MatchComplete)
-			includesCancellation = true
-			log.Infof("maker notification for cancel order received for order %s. match id = %s", oid, msgMatch.MatchID)
-			t.cancel.matches.maker = msgMatch
+		var qty uint64
+		if t.Type() == order.MarketOrderType && !trade.Sell {
+			qty = calc.BaseToQuote(msgMatch.Rate, msgMatch.Quantity)
 		} else {
-			includesTrades = true
-			t.matches[match.id] = match
+			qty = msgMatch.Quantity
+		}
+		newFill += qty
+
+		log.Infof("Starting negotiation for match %v for order %v with swap fee rate = %v, quantity = %v",
+			mid, oid, match.MetaMatch.Match.FeeRateSwap, qty)
+		match.SetStatus(order.NewlyMatched) // these must be new matches
+		newTrackers = append(newTrackers, match)
+	}
+
+	// Now that each Match in msgMatches has been validated, store them in the
+	// trackedTrade and the DB, and update each order's status and fill amount.
+	// First record any cancel order Match and update order status.
+	if cancelMatch != nil {
+		t.cancel.matches.maker = cancelMatch // taker is stored via processCancelMatch before negotiate
+		// Set the order status for both orders.
+		t.metaData.Status = order.OrderStatusCanceled
+		t.db.UpdateOrderStatus(t.cancel.ID(), order.OrderStatusExecuted)
+		// Store a completed maker cancel match in the DB.
+		makerCancelMeta := t.makeMetaMatch(cancelMatch)
+		makerCancelMeta.SetStatus(order.MatchComplete)
+		err := t.db.UpdateMatch(makerCancelMeta)
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
 		}
+		// Store a completed taker cancel match too.
+		takerCancelMeta := t.makeMetaMatch(t.cancel.matches.taker)
+		takerCancelMeta.SetStatus(order.MatchComplete)
+		err = t.db.UpdateMatch(takerCancelMeta)
+		if err != nil {
+			return fmt.Errorf("failed to update match in db: %w", err)
+		}
+	}
+	for _, match := range newTrackers {
+		t.matches[match.id] = match
 		err := t.db.UpdateMatch(&match.MetaMatch)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to update match in db: %w", err)
 		}
 	}
 
-	// Calculate the new filled value for the order.
-	isMarketBuy := t.Type() == order.MarketOrderType && !trade.Sell
-	var filled uint64
-	// If the order has been canceled, add that to the filled.
-	if t.cancel != nil && t.cancel.matches.maker != nil {
-		filled += t.cancel.matches.maker.Quantity
+	// NOTE: If Core was restarted, the loaded trackedTrades lack the cancel
+	// order Match data, preventing correct fill computation by summation of all
+	// t.matches quantities plus t.cancel.matches.maker.Quantity. For now, just
+	// increment the trade's FillAmt, which is stored and loaded with the order
+	// in the DB, by the newly-filled amount from these matches. Alternatively,
+	// a trade with known cancel could just be set to 100% filled without trying
+	// to do the accounting properly.
+	filled := trade.AddFill(newFill)
```

Comment:
The comment is outdated since all match data is loaded now. I personally prefer setting the fill based on the known matches, but I know that you mentioned that you liked this incremental approach. Got a good reason? Counting / setting based on match iteration is idempotent, so perhaps less error-prone.

Reply:
Right, the data is loaded now, just missing the cancel taker which isn't part of the sum.

My feeling on the incremental fill accounting is that losing track of a match won't break the sum, but I take your point that by doing SetFill with the sum you will not risk adding fill twice (idem-potency).  I'll switch back.  It's not a big issue to me.